### PR TITLE
refactor: using goovn bindings for adding ACL in addAllowACLFromNode()

### PR DESCRIPTION
**Pull request by vjayaramrh**
Signed-off-by: Vishwanath Jayaraman <vjayaram@redhat.com>

**- What this PR does and why is it needed**

This PR replaces the usage of `util.RunOVNNbctl()` function with the goovn bindings `ACLAdd` API for creating ACL rules.
Unit tests for the impacted function `addAllowACLFromNode()` has been added

**- Special notes for reviewers**

Note1. There are no instances of `acl-del` being invoked in the codebase
Note2. References to `acl-add` in the test code are removed as well per suggestion from @abhat .  

**- How to verify it**

**- Description for the changelog**

### CODE REVIEW

File: go-controller/pkg/ovn/policy_unit_test.go
Comment (abhat) [inline go-controller/pkg/ovn/policy_unit_test.go:97]: Since `addAllowACLFromNode` already takes the mock client as argument. It seems like we can follow the example in go_ovn_test.go and then you don't need the changes in go_ovn.go.

Comment (vjayaramrh) [reply go-controller/pkg/ovn/policy_unit_test.go:97]: Good catch, Thanks

File: go-controller/pkg/ovn/policy.go
Comment (abhat) [inline go-controller/pkg/ovn/policy.go:115]: might help to give more context to the generic command error by say logging similar to L112.

File: go-controller/pkg/testing/mock_stubs.go
Comment (abhat) [inline go-controller/pkg/testing/mock_stubs.go:62]: Convention is to move the methods that will actually be supported by the mock to their own file. Like for instance this will move the erstwhile stubbed code to a file called mock_acl.go.

File: go-controller/pkg/ovn/policy.go
Comment (vjayaramrh) [reply go-controller/pkg/ovn/policy.go:115]: Good suggestion, will add the context

File: go-controller/pkg/testing/mock_stubs.go
Comment (vjayaramrh) [reply go-controller/pkg/testing/mock_stubs.go:62]: OK, shall create a file called mock_acl.go

**Comment by abhat** [comment]
LGTM. @trozet @dcbw PTAL and merge if you are OK with the changes.

File: go-controller/pkg/ovn/policy.go
Comment (anfredette) [inline go-controller/pkg/ovn/policy.go:113]: Why do we want to ignore the ErrorExist error?  The lookup is only being done on the direction, match, and priority fields.  If the user was trying to create an ACL that contained different values for other fields (e.g., action), they wouldn't know that didn't happen.

**Review by anfredette — COMMENTED** [review]
@trozet made the following comment in an email:
"you need to becareful when you replace acl-adds with go-ovn bindings that you also at the same time replace delete and get. This is because go-ovn keeps an in memory cache of the OVN DB. The go-ovn bindings ensure that when you do an operation through the bindings that the cache is updated accordingly. However, if you circumvent go-ovn and use ovn-nbctl, the cache could become stale. We hit this doing this order of operations where we used add just for go-ovn but delete still used ovn-nbctl for pods: https://github.com/eBay/go-ovn/issues/104#issuecomment-685284445"

I didn't see any instances of acl-del or acl-list; however, there are several cases where ACLs are added and removed using the database commends.  Do we need to convert those to go-ovn bindings too before this is merged, or is this one an independent case?

File: go-controller/pkg/ovn/policy.go
Comment (vjayaramrh) [reply go-controller/pkg/ovn/policy.go:108]: Good suggestion, however, currently, the `defaultAllowPriority` is being used at https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/ovn/gress_policy.go#L321 and https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/ovn/policy.go#L227. When the functions that use  `defaultAllowPriority` get refactored, we can drop its definition as string constant and define it as integer constant instead, what do you think?

File: go-controller/pkg/ovn/policy.go
Comment (vjayaramrh) [reply go-controller/pkg/ovn/policy.go:113]: I am maintaining the equivalency of the original code, refer https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/ovn/policy.go#L104-L105 ,( note the `--may-exist` option)

File: go-controller/pkg/ovn/policy.go
Comment (vjayaramrh) [reply go-controller/pkg/ovn/policy.go:101]: @abhat @trozet Please weigh in with your comments as well, Thanks

Comment (trozet) [reply go-controller/pkg/ovn/policy.go:101]: iirc when @abhat first introduced using the go-ovn api we did have the clients as singletons, but then we decided that wasn't a great idea. I don't see a way other than passing it around to the functions that need it. @danwinship is probably the best person to bring into this conversation since he has a lot of expertise in golang code structure. What do you think Dan?

File: go-controller/pkg/ovn/policy.go
Comment (trozet) [reply go-controller/pkg/ovn/policy.go:113]: We usually allow for "may-exist" type of behavior because if ovnkube dies, when it restarts it will get all of the events from kubernetes and try to re-run all of this which may still already be present in the NBDB. So we try to keep things idempotent. We've had some discussion in the past about handling this better by maintaining and ovn-k8s state within CRDs or some other DB mechanism, but its kind of further down the road.

**Comment by trozet** [comment]
@vjayaramrh you mentioned:
Note1. There are no instances of acl-del being invoked in the codebase

so for addAllowACLFromNode we never call a delete for this if the node is deleted? Seems like a bug.

**Comment by anfredette** [comment]
Ok, so the assumption is that if we encounter a duplicate ACL, it's the one
we intended to be there.  Correct?

**Comment by anfredette** [comment]
The way OVN works, if the entity on which the ACL was added (logical switch
or port group) is deleted, the dependent ACLs get deleted also.  I was
assuming that we were counting on this behavior and not explicitly deleting
the ACLs.

File: go-controller/pkg/ovn/policy.go
Comment (anfredette) [reply go-controller/pkg/ovn/policy.go:108]: That's fine with me.

File: go-controller/pkg/ovn/policy.go
Comment (danwinship) [reply go-controller/pkg/ovn/policy.go:101]: > iirc when @abhat first introduced using the go-ovn api we did have the clients as singletons, but then we decided that wasn't a great idea.

Hm... I don't remember that.

`pkg/config` and `pkg/util/ovs.go` use singletons (which can be overridden for unit tests) and that seems to work ok?

> could cause difficulties if the same code needs to manipulate multiple NB and SB databases.

but I don't think we'd ever need to do that.


From all the code I've seen in kubernetes, I don't think anyone has figured out an optimal way to do this. All the solutions are ugly in their own way.

Comment (abhat) [reply go-controller/pkg/ovn/policy.go:101]: The comment was I think from @dcbw. The passing of go-client as arguments is valuable because in future if we decide to slice and dice the ovn-controller (across namespaces, albeit hypothetically, for instance), you could like @anfredette pointed out create multiple nbdb/sbdb client instances one for each controller shard.

So while this is inconvenient and one may argue it is premature optimization at this point, we should try to keep that abstraction.

File: go-controller/pkg/ovn/policy.go
Comment (anfredette) [inline go-controller/pkg/ovn/policy.go:120]: If the ACL already exists, you'll get the "ErrorExist" error from the ACLAdd call.  And, if there's an error, the aclcmd returned is nil.  If you then call Execute with a nil command, it's treated as a no-op and will execute without error.

I think you need to move the check for ErrorExist above the check for "err == nil".

**Review by anfredette — CHANGES_REQUESTED** [review]

File: go-controller/pkg/ovn/policy_unit_test.go
Comment (anfredette) [inline go-controller/pkg/ovn/policy_unit_test.go:42]: I think you should expect the goovn.ErrorExist error here on the ACLAdd.

Comment (vjayaramrh) [reply go-controller/pkg/ovn/policy_unit_test.go:42]: @anfredette thanks, let me look into it

File: go-controller/pkg/ovn/policy.go
Comment (vjayaramrh) [reply go-controller/pkg/ovn/policy.go:120]: @anfredette good catch, Thanks

**Review by anfredette — APPROVED** [review]
LGTM w/latest changes.  Thanks.

File: go-controller/pkg/ovn/policy.go
Comment (trozet) [reply go-controller/pkg/ovn/policy.go:125]: @anfredette ah I see what you mean. Sorry I missed it. I presumed the may-exist would also check action for ACL equivalency. Should this be something fixed in go-ovn or OVN? For now should we find the ACL from the go-ovn cache and compare the action as a workaround to be 100% sure it is the same as what we are intending?

Comment (anfredette) [reply go-controller/pkg/ovn/policy.go:135]: If we want to be 100% sure, then I think that's what we'd need to do.  However, if we know we always create them a certain way, then maybe we could skip the comparison.  That's not the way the APIs work currently, and I don't think Han would be ok with changing the semantics.  Perhaps an option is to add a new behavior, like --may-exist, but we could call it something else like --force, which would change the ACL to have all the properties in the create.  The new ACL APIs I added haven't been there that long.  Han may be okay with updating them.

Comment (anfredette) [reply go-controller/pkg/ovn/policy.go:135]: I think the way the new API could work is:
- if the ACL isn't found, just create it as normal.
- if the ACL is found, then compare the rest of the fields and update any that are different.

Comment (trozet) [reply go-controller/pkg/ovn/policy.go:135]: Seems reasonable to me. For now we can push this PR through as the chance of this happening with the changes in this patch is pretty much none. Let's see how the discussion in go-ovn goes.

**Review by trozet — APPROVED** [review]
/lgtm
waiting on CI

File: go-controller/pkg/ovn/policy.go
Comment (anfredette) [reply go-controller/pkg/ovn/policy.go:135]: I'd have to double-check, but I think the current code using RunOVNNbctl would have the same issue.

**Pull request merged**
